src/hb-ot-cmap-table.hh

```diff
@@ -1215,13 +1361,43 @@ struct cmap
     for (const EncodingRecord& _ : encodingrec_iter)
     {
       hb_set_t unicodes_set;
-      (base+_.subtable).collect_unicodes (&unicodes_set);
+      hb_map_t cp_glyphid_map;
+      (base+_.subtable).collect_mapping (&unicodes_set, &cp_glyphid_map);
 
       unsigned format = (base+_.subtable).u.format;
-
-      if (format == 4) c->copy (_, + it | hb_filter (unicodes_set, hb_first), 4u, base, plan, &format4objidx);
-      else if (format == 12) c->copy (_, + it | hb_filter (unicodes_set, hb_first), 12u, base, plan, &format12objidx);
-      else if (format == 14) c->copy (_, it, 14u, base, plan, &format14objidx);
+      if (!plan->glyphs_requested->is_empty ())
```

Comment:
Can we keep this if-else code for efficiency reasons?
I tried consolidating two approaches but bots failed due to timeout in cmap.tests. This is because:
when "--gids" option is not used, the other approach is more efficient: collect_mapping, iterate "input unicodes only" which is much faster than iterating all codepoints in a subtable and then filter out codepoints that don't belong to that subtable. 

Font file used in cmap.tests is large, taking several minutes to convert into ttx file. In this special case, if "--gids" option is not used, iterating input unicodes is significantly faster.

Reply:
Ok in that case we can keep the alternate path, but can you add a comment to the else which explains why we have the alternate implementation for that case (ie. for efficiency)